### PR TITLE
Add rabbitmq tls metadata

### PR DESCRIPTION
**Pull request by SpikeWong**
Signed-off-by: spike <hello@spike.wiki>

# Description

Add TLS related metadata configuration to rabbitmq

## Issue reference

 #2008
 #2246 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#2953

### CODE REVIEW

File: pubsub/rabbitmq/metadata.go [hunk 1]
```diff
@@ -97,6 +101,14 @@ func createMetadata(pubSubMetadata pubsub.Metadata, log logger.Logger) (*metadat
...
+		uri, err := amqp.ParseURI(result.connectionString)
+		if err != nil {
+			return &result, fmt.Errorf("%s invalid connection string: %s, err: %w", errorMessagePrefix, result.connectionString, err)
+		}
+		result.protocol = uri.Scheme
+	}
+
 	if val, found := pubSubMetadata.Properties[metadataProtocolKey]; found && val != "" {
 		result.protocol = val
 	}
```
Comment (anoymouscoder) [inline pubsub/rabbitmq/metadata.go:114]: Not sure if I need to check that the protocol parsed from the uri is the same as the configured protocol. If not, raised an error?

Comment (berndverst) [reply pubsub/rabbitmq/metadata.go:114]: I don't think that's necessary. The protocol metadata key can be an explicit override for users who know what protocol they want to use.

**Review by berndverst — CHANGES_REQUESTED** [review]
Thanks for the PR!

Please address the linter issues. (See build Linux_amd64 action run)

**Comment by SpikeWong** [comment]
> Can you please add a unit test for the following cases:
> 
> * Connection URI set (with protocol) & protocol is not set -> protocol is correctly inferred
> * Connection URI with missing protocol part & protocol is not set -> some error (or default?)
> * Connection URI (with protocol) & protocol is set to something else -> the explicit protocol value is used

I'm not sure if we should override the protocol in the connectionString when both the connectionString and the protocol are set.
According to this func, the corresponding value in connectionString is currently not overwritten if username, host and connectionString are set at the same time.
So if we were to do this with the protocol, would we also change the behaviour of the other fields https://github.com/dapr/components-contrib/blob/9870c5e33beb8706a6e76e5c95d55a39b40d7fa3/pubsub/rabbitmq/metadata.go#L233-L250

**Comment by berndverst** [comment]
@SpikeWong ok, in that case you need to compare the protocol between `connectionString` and `protocol` metadata and throw an error if these mismatch.

**Comment by SpikeWong** [comment]
> @SpikeWong ok, in that case you need to compare the protocol between `connectionString` and `protocol` metadata and throw an error if these mismatch.

finished

File: pubsub/rabbitmq/metadata.go [hunk 2]
```diff
@@ -203,6 +218,11 @@ func createMetadata(pubSubMetadata pubsub.Metadata, log logger.Logger) (*metadat
 		result.defaultQueueTTL = &ttl
 	}
 
+	result.TLSProperties, err = pubsub.TLS(pubSubMetadata.Properties)
+	if err != nil {
+		return &result, fmt.Errorf("%s invalid TLS configuration: %w", errorMessagePrefix, err)
+	}
+
```
Comment (Taction) [inline pubsub/rabbitmq/metadata.go:225]: If the protocol is AMQPS, check if TLS Properties are provided. If not, return the error info to inform users that the tls config is missed for amqps protocol.

Comment (SpikeWong) [reply pubsub/rabbitmq/metadata.go:225]: Based on https://www.rabbitmq.com/ssl.html#enabling-tls, if rabbitmq server configure `ssl_options.verify` as `verify_none` or `ssl_options.fail_if_no_peer_cert` as `false`, the client is allowed to use amqps protocol without certificates.

**Review by yaron2 — APPROVED** [review]

**Comment by yaron2** [comment]
LGTM. @berndverst see if you're happy.

**Review by berndverst — APPROVED** [review]

**Pull request merged**